### PR TITLE
Add osu!taiko "Hidden" mod

### LINKED ISSUE #8548: Implement Hidden for osu!taiko

### PR DESCRIPTION
**Pull request by chayleaf**
https://user-images.githubusercontent.com/191335/122595061-d6aeee80-d0a2-11eb-9ab2-9872e74193c0.mp4

This PR adds HD mod to Taiko, I got it as close to stable as I could.

In stable, HD has smaller visible area. To account for that, I added a mod that changes visibility.

It also fixes the complaints about screen size being fixed from some players, but ensures consistent replay experience. Ideally, the mod needs to become a system mod like touchscreen, configurable from game settings - but I'm not sure about what to do with the HD mod, as visible area used to be fixed to 4:3 with HD mod.

Closes https://github.com/ppy/osu/issues/8548

Edit: Just realized "4:3" as the description of the playfield size correspondent with the osu!stable 4:3 area isn't descriptive to most people, should I remove the special case from SettingDescription?

Edit 2: Lints didn't work cause I didn't have .Net Core 3.1 installed, will install it and fix the issues.

### CODE REVIEW

File: osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs:55]: As far as I can see this is copied verbatim from `OverlappingScrollAlgorithm`. It should be shared somehow, possibly by subclassing `SortedList<MultiplierControlPoint>` and adding this there as an instance method.

File: osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs:30]: Need explanation for this. What happens if it isn't used with max playfield size?

File: osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs
Comment (bdach) [inline osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs:16]: Needs to be named `AspectRatio`, made a property, and documented with xmldoc.

File: osu.Game/Rulesets/UI/DrawableRuleset.cs
Comment (bdach) [inline osu.Game/Rulesets/UI/DrawableRuleset.cs:68]: Unsure why `virtual`. Don't really believe this should be overridable, `CreatePlayfieldAdjustmentContainer()` should suffice for a customisation API surface.

File: osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs:70]: As far as I can see, stable code for this is
```csharp
protected override int LocalPreEmpt
{
    get
    {
        float multiplier = 1;

        if (Player.currentScore?.Processor != null && ModManager.CheckActive(hitObjectManager.ActiveMods, Mods.HardRock))
            multiplier = 1.4f;

        return (int)(600 / hitObjectManager.SliderVelocityAt(StartTime) / 0.6f * 1000 / multiplier);
    }
}
```
which is then used in
```csharp
Transformation fadeout = hidden ?
    new Transformation(TransformationType.Fade, 1, 0, StartTime - (int)(LocalPreEmpt * 0.6), StartTime - (int)(LocalPreEmpt * 0.3)) { TagNumeric = ARMED } :
    new Transformation(TransformationType.Fade, 1, 0, StartTime + hitObjectManager.HitWindow100, StartTime + hitObjectManager.HitWindow50) { TagNumeric = ARMED };
```

File: osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs:34]: How was this derived?

File: osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs:28]: This needs to be explained better. A value between 0.5 and 2 with no units doesn't tell much. Especially so that it's apparently a ratio relative to 16:9 which is extra confusing.

**Review by bdach — DISMISSED** [review]

File: osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs:30]: If it isn't used with max playfield size, no notes are visible. Should it be scaled with playfield size?

File: osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs:34]: Manually. osu!lazer's playfield has more notes to the left of the hit area, and that seems to make the equivalent to 4:3 0.7 rather than 0.75. Maybe I'm wrong about the reasoning, but either way, 0.75 showed too many notes, and 0.7 *seems* just right.

Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs:34]: Manually. osu!lazer's playfield has more notes to the left of the hit area, and that seems to make the equivalent to 4:3 0.7 rather than 0.75. Maybe I'm wrong about the reasoning, but either way, 0.75 showed too many notes, and 0.7 *seems* just right.

File: osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs:28]: Right, this is a poor description

**Comment by bdach** [comment]
In general I'm not really sure I follow all of the aspect ratio concerns. Does the `TaikoPlayfieldAdjustmentContainer` already not account for/normalise aspect ratio to 16:9 correctly?

**Comment by pavlukivan** [comment]
the problem is stable HD ratio is equivalent to 4:3, not to mention some people prefer 4:3 in general (I'm not one of them, but it's the reason I made it a separate mod as opposed to part of the HD mod)

File: osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs:70]: I sadly can't see how `600 / 0.6 * 1000` translates to `15000`, which is why I gave up on finding the precise value and chose it manually instead

File: osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
Comment (bdach) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs:30]: Not sure, but if it's coupled somehow to the other mod, then the ideal state of things would be that one mod's settings influence the other's allowable bounds to prevent brokenness. However this is unprecedented in lazer and we don't really have a good way of doing that sort of stuff.

Incidentally taiko HD and HR also seem to be somehow intertwined on stable, [which already came up once](https://github.com/ppy/osu/pull/11364#discussion_r550181930) but ended up being unresolved for now.

**Comment by bdach** [comment]
I think in that case the way I'd go forward if I was to suggest one is to add a ruleset-specific setting (in the game settings) for that, with only 4:3 and 16:9 options for now, rather than add this almost-infinite tinkering with the aspect ratio. If that becomes a setting then the HD setting can be customisable better based on the aspect ratio chosen by the user. I'd probably split all of that to a separate PR too to avoid mixing concerns.

That said I'm not super privy to taiko mechanics with respect to aspect ratio and haven't thought about it much, so I'd probably appreciate backing from someone from the team in order to determine if that is a proper way to go.

**Comment by pavlukivan** [comment]
In stable, many players do use different resolutions, some use 5:4, some use 21:9, some use 16:6. As a taiko player, I do think it makes sense to add custom playfield size support.

HD is always 4:3 though - and it's not a vanity issue, if it was unlimited 16:9 HD would be way easier, giving you much more time to recognize an object. I don't think making HD aspect ratio user-configurable without any indication of how exactly it's configured is a good solution.

**Comment by pavlukivan** [comment]
So what you're saying is HD should force-override the playfield size to 4:3 and disable any customization?

**Comment by bdach** [comment]
I guess if the playfield adjustment is visible among the mods with which a score was achieved, then that can be used to distinguish normal HD plays from potentially "cheesed" HD plays via adjusting the ratio...? So maybe not on that one.

**Comment by pavlukivan** [comment]
So the 2 mods shouldn't be incompatible then. I'll make the adjustments soon, I guess I should still override the default HD ratio to 4:3 though because it'd be annoying to enable the playfield resize mod along with hd manually every time.

**Comment by bdach** [comment]
All the above notwithstanding I still believe that splitting off the aspect ratio adjustment to a separate pull and just assuming constant 16:9 ratio for HD in this one may be beneficial for review/merge speed.

File: osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs:35]: removed this mod for now

File: osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModPlayfieldResize.cs]: i actually calculated this properly now!

**Comment by bdach** [comment]
Blocking this, pending investigation of lifetime issues mentioned in https://github.com/ppy/osu/commit/9bd8971ed95dd44f10ad878c644cdc87abed5cfc. Will see if I can figure out what's going on.

**Comment by pavlukivan** [comment]
https://osu.ppy.sh/b/1558762 with lifetimes set + autoplay, on this map the first note gets hit, the 2nd and 3rd ones get hit with a 100, the 4th one gets missed. With opacity high enough 100s/misses don't happen.

File: osu.Game.Rulesets.Taiko/TaikoRuleset.cs
Comment (peppy) [inline osu.Game.Rulesets.Taiko/TaikoRuleset.cs:152]: Revert

File: osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs
Comment (peppy) [inline osu.Game.Rulesets.Taiko/UI/TaikoPlayfieldAdjustmentContainer.cs:26]: revert

File: osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs
Comment (peppy) [inline osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs:81]: This can't be public in its current state. If it is required for some reason or another I think it needs to be in its own isolated PR.

This PR should *only* touch the `TaikoModHidden` file, nothing else.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs
Comment (peppy) [reply osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs:81]: To rephrase before you add all these changes back: I don't think any of this should be required in the first place, and this mod should be using `IApplicableToBeatmap` and then adjusting via `ControlPointInfo`.

Comment (pavlukivan) [reply osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs:81]: That would require a lot of code duplication, is that alright?

Comment (peppy) [reply osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs:81]: Which code and how much?

Comment (pavlukivan) [reply osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs:81]: all the code to generate MultiplierControlPoint's from ControlPointInfo

Comment (peppy) [reply osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs:81]: I think it will be easier to write this from scratch myself rather than attempt to work with this PR. See how you go, I think it should be possible. I strongly believe there's no need to consume the post-processed control points.

Comment (pavlukivan) [reply osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs:81]: you were right, a lot of MultiplierControlPoint's complexity is coming from mania

File: osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
Comment (peppy) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs:26]: Rather than putting `9/13` in here you should probably put the full `((4 / 3) / (16 / 9))` or whatever the equation is (there's no real reason to simplify the equation out when the more verbose version actually reveals useful information).

**Comment by bdach** [comment]
Have applied a few basic changes, including replacing the lifetime workaround. Using `GetEndTime() + HitWindowFor(HitResult.Miss)` was almost correct, but I forgot that `ApplyCustomUpdateState` also runs for `ArmedState.{Hit,Miss}`, and in those cases prolonging the object's lifetime past the judgement time is *not* wanted. Also added a rudimentary test scene for the mod because might as well.

Still not sure about that preempt time calculation, would have to debug stable a bit to see how that works exactly. But I suppose I'm going to unblock & unassign for now pending further review.

**Comment by pavlukivan** [comment]
Perhaps preempt needs to be 10000 / hd_sv_scale as opposed to 10000 * 1.4? I do think it requires debugging in stable.

**Comment by bdach** [comment]
I've gone to check stable, and I believe that the final formula should be

```csharp
var preempt = 10 / difficulty.SliderMultiplier * (controlPointInfo.TimingPointAt(position)?.BeatLength ?? DEFAULT_BEAT_LENGTH);
```

My working is listed in the collapsible box below:

<details>
<summary>Stable math</summary>

```
Stable calculations go as follows:
----------------------------------

LocalPreEmpt = (600 / SliderVelocityAt(time) / 0.6 * 1000 / multiplier)
             // Rearrange and merge constants.
             = 1_000_000 / SliderVelocityAt(time) / multiplier
             // Multiplier is 1.4 iff HR is active, otherwise it is 1. Skip for now.
             = 1_000_000 / SliderVelocityAt(time)

SliderVelocityAt(time) = SliderScoringPointDistance * DifficultySliderTickRate * (1000 / beatLength)

SliderScoringPointDistance = 100 * DifficultySliderMultiplier / DifficultySliderTickRate

SliderVelocityAt(time) = SliderScoringPointDistance * DifficultySliderTickRate * (1000 / beatLength)
                       // Substitute in SliderScoringPointDistance.
                       = 100 * DifficultySliderMultiplier / DifficultySliderTickRate * DifficultySliderTickRate * (1000 / beatLength)
                       // Eliminate terms which cancel out.
                       = 100 * DifficultySliderMultiplier * (1000 / beatLength)
                       // Rearrange and merge constants.
                       = 100_000 * DifficultySliderMultiplier / beatLength

LocalPreEmpt = 1_000_000 / SliderVelocityAt(time)
             // Substitute in SliderVelocityAt(time).
             = 1_000_000 / 100_000 / DifficultySliderMultiplier * beatLength
             // Simplify constants.
             = 10 / DifficultySliderMultiplier * beatLength
```
</details>

That makes two things in particular:

* Besides the formula being off wrt constants (which was to be expected), I couldn't see any way in which overall difficulty would HD pre-empt time, so it does seem to me that that should be removed. I think I've looked through all usages in stable and I can't seem to find a place where it would apply to that. So uless @pavlukivan is sure that OD does influence pre-empt time it does seem that that should be removed.
* The second part is the aspect ratio stuff. The above calculations all hold up for 4:3. This PR attempts to compensate for lazer currently being hard-locked to 16:9 by increasing the slider multiplier proportionally to the difference between playfield length at 16:9 to 4:3, but that also makes the objects scroll faster, because they have to cover more distance in the same time.
 
  An alternative that would preserve scroll time would be to just increase the pre-empt time on its own without touching the multiplier, but that in turn causes players to have to memorise more notes. I don't have a strong opinion on this matter, as it's a game balance one and I don't play taiko well enough to know how it matters.

**Comment by pavlukivan** [comment]
EZ/HR mods decrease/increase base SV, not just OD. OD doesn't matter indeed, and I never said it does.

**Comment by bdach** [comment]
I see, thanks for clarifying.

**Comment by bdach** [comment]
I have applied further changes to this pull and don't see any more left to apply. For an initial implementation I *think* this should work well enough. The HR crosstalk, as well as the aspect ratio stuff, can be tackled later down the line.

File: osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs
Comment (peppy) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModHidden.cs:30]: Removing this setting for now. If it's added, it should be added in `ModHidden` (the base class) and implemented for all bundled rulesets.

**Review by peppy — APPROVED** [review]

**Review by bdach — APPROVED** [review]

**Pull request merged**
